### ISSUE TITLE
Support for multiple inputs

### ISSUE THREAD

**Issue by r-marques** [bigchaindb/bigchaindb#128]
Ability to transfer multiple digital assets in a single transaction.
The transaction structure will change with this feature.

The current transaction looks like:
```json
{
    "id": "<sha3 hash>",
    "transaction": {
        "current_owner": "<pub-key>",
        "new_owner": "<pub-key>",
        "input": "<sha3 hash>",
        "operation": "<string>",
        "timestamp": "<timestamp from client>",
        "data": {
            "hash": "<SHA3-256 hash hexdigest of payload>",
            "payload": {
                "title": "The Winds of Plast",
                "creator": "Johnathan Plunkett",
                "IPFS_key": "QmfQ5QAjvg4GtA3wg3adpnDJug8ktA1BxurVqBD8rtgVjP"
            }
        }
    },
    "signature": "<signature of the transaction>"
}
```

The new transaction structure will need to have a inputs and outputs:
```json
{
    "id": "<sha3 hash>",
    "transaction": {
        "current_owner": "<pub-key>",
        "new_owner": "<pub-key>",
        "inputs": [
            {
                "txid": "<sha3 hash>",
                "number": "input number"
            }
            ],
        "operation": "<string>",
        "timestamp": "<timestamp from client>",
        "data": {
            "hash": "<SHA3-256 hash hexdigest of payload>",
            "payload": {
                "title": "The Winds of Plast",
                "creator": "Johnathan Plunkett",
                "IPFS_key": "QmfQ5QAjvg4GtA3wg3adpnDJug8ktA1BxurVqBD8rtgVjP"
            }
        }
    },
    "signature": "<signature of the transaction>"
}
```

At this point I am not sure if we need to have outputs. If we add multiple outputs to possibly different owners we also need a way to link the inputs to the outputs for non divisible assets.

Its probably best to support multi outputs when addressing divisible assets and/or crypto conditions (where we will need to add a fulfillment to the inputs and a condition to the outputs) 

Right now I will focus on adding only multiple input support to create transactions between a single `current_owner` and a single `new_owner`.

**Comment by r-marques** [comment]
```json
{
    "id": "<sha3 hash>",
    "transaction": {
        "fulfillments": [
                {
                    "current_owners": ["list of <pub-keys>"],
                    "input": {
                        "txid": "<sha3 hash>",
                        "cid": "condition index"
                    },
                    "fulfillment": "fulfillement of condition cid",
                    "fid": "fulfillment index"
                }
            ],
        "conditions": [
                {
                    "new_owners": ["list of <pub-keys>"],
                    "condition": "condition to be met",
                    "cid": "condition index (1-to-1 mapping with fid)"
                }
            ],
        "operation": "<string>",
        "timestamp": "<timestamp from client>",
        "data": {
            "hash": "<SHA3-256 hash hexdigest of payload>",
            "payload": {
                "title": "The Winds of Plast",
                "creator": "Johnathan Plunkett",
                "IPFS_key": "QmfQ5QAjvg4GtA3wg3adpnDJug8ktA1BxurVqBD8rtgVjP"
            }
        }
    },
}
```

Initial implementation following the [interledger specs](https://interledger.org/five-bells-condition/spec.html) for crypto conditions.

##### Fulfillments
Fulfillments are a list of inputs. They contain a pointer to the condition that needs to be fulfilled in order to *spend* the input.

- `current_owners`: Public key (or list of public keys in case of multi/threshold sigs) that will be used to validate the fulfillment.
- `input`: Pointer to the digital asset to *spend* represent by a condition to fulfill. It contains the `txid` of the transaction and `cid` the condition index (similar to *vout* in bitcoin).
- `fulfillment`: the actual fulfillment of the condition. It is a signature of the `input`, `condition` (from the previous transaction), and the rest of the data in the transaction (`operation`, `timestamp`, `data`, `conditions`)
- `fid`: fulfillment index so that we can reference it when defining the conditions (we need to define a condition for each input used)

To validate if the signature in the `fulfillment` is correct we need to make sure that that the `current_owners` pub key(s) verifies the the signature of the message (which is a document containing the fields described above) 

##### Conditions
Conditions are a list of outputs. They represent a condition that needs to be fulfilled by the current owner to _spend_ the digital asset

- `new_owners`: new owner(s) of the digital asset
- `condition`: The pub key(s) that will need to satisfy a fulfillment (same as new_owners but using the interledger format)
- `cid`: condition index. It specifies to which input this condition applies

**Comment by diminator** [comment]
solved with #174
